Issue 4 inheritance problems

pbtura opened the pull request:
Fix for issue where dto generation was not working properly for classes with inheritance. Currently, the DtoConfig class uses the declaringClass for the getter method to determine if a property should be included. Need to determine if the setter method should be checked as well.

Commit:
Added additional classes to test inheritance

Commit:
Fix for generated constructor with inheritance

Fixes an issue where the constructor for DTOs with inheritance would
contain duplicate parameters.

Commit:
Added code to exclude inherited properties from mapping

Modified the addPropertiesFromDomainObject() method of DtoConfig.
Inherited properties will now be excluded if the Dto is a child class.

Commit:
Added codegen for the GreenAccountDto

stephenh commented on dtonator/src/main/java/com/bizo/dtonator/GenerateDto.java:
```diff
@@ -126,8 +132,13 @@ private void addCopyOfMethod() {
-      // first make copies of children if needed
-      for (final DtoProperty dp : c.getAllProperties()) {
+
+      List<DtoProperty> properties = new ArrayList<DtoProperty>();
```
Why this change?

pbtura replied:
This looks like a leftover change from something else I tried out and discarded. Go ahead and roll this back if you want.

stephenh commented on dtonator/src/main/java/com/bizo/dtonator/config/DtoConfig.java:
```diff
@@ -64,7 +64,7 @@ public DtoConfig(final TypeOracle oracle, final RootConfig root, final String si
 
   public List<DtoProperty> getAllProperties() {
-    return list(getInheritedProperties()).with(getProperties());
+    return list(getInheritedProperties()).with(getProperties()).unique();
```
Do you still need this given that AFAIU the properties should not be duplicated in the first place anymore?

pbtura replied:
It looks like you are correct. My original attempt still required the unique() call but this version does not. I tried taking out the unique() call and the equals()/ hashcode() and everything still works correctly.

stephenh commented on dtonator/src/main/java/com/bizo/dtonator/config/DtoProperty.java:
```diff
@@ -147,4 +147,59 @@ public boolean isChainedId() {
   }
 
+  @Override
+  public int hashCode() {
```
Similar, do we still need these? If so, I'd generally prefer the apache commons HashCodeBuilder/EqualsBuilder as these Eclipse-/Intellij-generated methods are painful to read.

stephenh commented on dtonator/src/main/java/com/bizo/dtonator/properties/ReflectionTypeOracle.java:
```diff
@@ -23,7 +23,9 @@
         pd.getWriteMethod() == null,
         pd.getReadMethod() == null ? null : pd.getReadMethod().getName(),
-        pd.getWriteMethod() == null ? null : pd.getWriteMethod().getName()));
+        pd.getWriteMethod() == null ? null : pd.getWriteMethod().getName(),
```
I'm curious, if instead of tracking these declaring class properties, only to ignore them later, can we just not return them from `ReflectionTypeOracle.getProperties` at all?

E.g. right now `getProperties` returns `Props` in both the class + all sub-classes.

If we change that to only return `Props` in just the class itself, would that solve the problem of duplicate `Props` in a potentially cleaner way?

pbtura replied:
If I understand correctly, you are suggesting doing the filtering in the ReflectionTypeOracle.getProperties() call rather than in DtoConfig right? That could work, but it would require passing in an additional parameter to tell the oracle getProperties() method whether or not the parent class properties should be filtered. I can give this a try and see if I run into any difficulties.

stephenh commented on dtonator/src/main/java/com/bizo/dtonator/properties/Prop.java:
```diff
@@ -6,21 +6,43 @@
   public final boolean readOnly;
-  public final String getterMethodName;
-  public final String setterNameMethod;
+  private final String getterMethodName;
```
Given this is just an internal DTO and not a public API, I generally prefer the sacrilege of just using public final fields. Especially since some of the fields work that way already, it'd be better to be consistent that have some public fields and some private fields+getters.

pbtura replied:
You know the way these fields are meant to be used, and now I have a pretty good idea, but someone else who forks this code later might not. Making the fields private and read-only makes it crystal clear that nothing should change those properties once they are set. 

The best way to do it would actually be to store the Method itself and make the getters access the properties of the method. Unfortunately, that would make testing really hairy, so I don't think that is an option.

stephenh reviewed (changes requested):
Hey, sorry it took awhile, but made some comments. Thanks!

Commit:
Rolled back some unneeded changes

Commit:
Moved the property filtering into the oracle

As requested in the PR comments, moved the filtering of inherited
properties from the DtoConfig into the ReflectionTypeOracle. With this
change, the declaring class no longer needs to be saved in Prop.

Commit:
Added joist to the ivy dependencies for dtonator-features.
